src/root/async.c

```diff
@@ -109,7 +114,8 @@ unsigned __stdcall startthread(void *p)
     AsyncRead *aw = (AsyncRead *)p;
 
     //printf("aw->filesdim = %p %d\n", aw, aw->filesdim);
-    for (size_t i = 0; i < aw->filesdim; i++)
+    size_t dim = aw->filesdim;
+    for (size_t i = 0; i < dim; i++)
```

Comment:
There are two issues that this PR fixes:

The main thread [calls][1] `aw->start()` (`aw` is an instance of [`AsyncRead`][2]), which kicks off a background thread which starts reading files asynchronously.

The main thread could call `AsyncRead::dispose(aw)` while the background thread is still working (i.e. it is still inside the loop in [`startthread(..)`][3]

The first problem is that before this patch the _WIN32 version of [`AsyncRead::dispose(..)`][4] didn't wait for the background thread to finish and freed the AsyncRead instance, which could crash the background thread inside the [`startthread(..)`][3]. This is what [this][5] change fixes.

The other problem is that it was possible just after the last call to [`SetEvent()`][6] that [`AsyncRead::dispose()`][7] finishes (because all files marked as processed) and frees `aw`, and then the background thread continues and tries to access [`aw->filesdim`][8] for last time and crashes since `aw` is already freed.

My last [comment][9] was incorrect because I didn't see the problem, but Ketmar explained it to me in the [bugzilla issue][10] and now I agree with him.

I hope this makes things clear now.

[1]: https://github.com/D-Programming-Language/dmd/blob/973b7f88c0aa06fcb01334d406544f88aae22238/src/mars.c#L1324

[2]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.h#L23

[3]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L113

[4]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L102

[5]: https://github.com/D-Programming-Language/dmd/pull/4363/files#diff-d15ca38e1ad1fb0a42535fbb6551a62aR102

[6]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L116

[7]: https://github.com/D-Programming-Language/dmd/pull/4363/files#diff-d15ca38e1ad1fb0a42535fbb6551a62aR105

[8]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L112

[9]: https://github.com/D-Programming-Language/dmd/pull/4363#discussion_r24554262

[10]: https://issues.dlang.org/show_bug.cgi?id=14097#c5